messenger/message-recorder.rst

```diff
@@ -0,0 +1,120 @@
+This middleware enables us to add messages to a separate transaction that will only be
+dispatched *after* the current message handler finishes.
+
+Referencing the above example, this means that the ``UserSignedUp`` event would not be handled
+until *after* the ``SignUpUserHandler`` had completed and the new ``User`` was persisted to the
+database. If the ``SignUpUserHandler`` encounters an exception, the ``UserSignedUp`` event will
+never be handled and if an exception is thrown while sending the welcome email, the Doctrine
+transaction will not be rolled back.
+
+To enable this, you need to add the ``handle_message_in_new_transaction``
+middleware. Make sure it is registered before ``DoctrineTransactionMiddleware``
+in the middleware chain.
+
+**Note:** The ``handle_message_in_new_transaction`` middleware must be loaded for *all* of the
+buses. For the example, the middleware must be loaded for both the command and event buses.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/messenger.yaml
+        framework:
+            messenger:
+                default_bus: messenger.bus.command
+
+                buses:
+                    messenger.bus.command:
+                        middleware:
+                            - validation
+                            - handle_message_in_new_transaction
+                            - doctrine_transaction
+                    messenger.bus.event:
+                        default_middleware: allow_no_handlers
+                        middleware:
+                            - validation
+                            - handle_message_in_new_transaction
+                            - doctrine_transaction
+
+
+.. code-block:: php
+
+    namespace App\Messenger\CommandHandler;
+
+    use App\Entity\User;
+    use App\Messenger\Command\SignUpUser;
+    use App\Messenger\Event\UserSignedUp;
+    use Doctrine\ORM\EntityManagerInterface;
+    use Symfony\Component\Messenger\Envelope;
+    use Symfony\Component\Messenger\Stamp\Transaction;
+    use Symfony\Component\Messenger\MessageBusInterface;
+
+    class SignUpUserHandler
+    {
+        private $em;
+        private $eventBus;
+
+        public function __construct(MessageBusInterface $eventBus, EntityManagerInterface $em)
+        {
+            $this->eventBus = $eventBus;
+            $this->em = $em;
+        }
+
+        public function __invoke(SignUpUser $command)
+        {
+            $user = new User($command->getUuid(), $command->getName(), $command->getEmail());
+            $this->em->persist($user);
+
+            // The Transaction stamp marks the event message to be handled
+            // only if this handler does not throw an exception.
+
+            $event = new UserSignedUp($command->getUuid());
+            $this->eventBus->dispatch(
+                (new Envelope($event))
+                    ->with(new Transaction())
+            );
```

Comment:
This is a bit confusing: why/when would this handler throw an exception? This is the last line of this function... so if we get there, how would an exception occur? And where should the flush live?

Reply:
Indeed, at this point is should be okay... unless there is a middleware doing so crazy things 😉